screencaptor/src/main/java/com/wealthfront/screencaptor/viewmutator/ViewMutator.kt

```diff
@@ -0,0 +1,66 @@
+package com.wealthfront.screencaptor.viewmutator
+
+import android.view.View
+import androidx.annotation.IdRes
+import androidx.test.espresso.UiController
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.matcher.ViewMatchers
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+
+abstract class ViewMutator<S : View, T>(private val viewClass: Class<S>, private val desiredValue: T) {
+  @IdRes
+  protected abstract fun key(): Int
+
+  protected abstract fun originalViewState(view: S): T
+
+  fun mutate(view: S) {
+    view.setTag(key(), originalViewState(view))
+    mutateView(view, desiredValue)
+  }
+
+  fun restore(view: S) {
+    val state = storedViewState(view)
+    restoreView(view, state)
+    view.setTag(key(), null)
+  }
+
+  protected abstract fun mutateView(view: S, value: T)
+
+  protected abstract fun restoreView(view: S, value: T)
+
+  @Suppress("UNCHECKED_CAST")
+  private fun storedViewState(view: S): T {
+    return view.getTag(key()) as T
+  }
+
+  fun getRestorationViewMatcher(): Matcher<View> {
```

Comment:
TODO kust use original matcher. Otherwise, multiple mutators using the same key will trigger ambigious matcher. Original matcher must be unambigious anyway (edge case: mutator actions interfere with that?)